Feat: Auto-detect timeout evasion

hasanlatch007 opened the pull request:
This should close #370

hasanlatch007 commented:
Testing failed because `timedOutUsers` does not exists in the db yet.

geisterfurz007 reviewed (changes requested):
We shouldn't inline so much code into events and other programs, especially given the new event system which tries to shift focus onto more capsulated features.

Instead of putting some bits into moderator.ts, guild-member-update.ts, member-leave.ts, message.ts *and* where-are-you-from.ts, you should use the new event system for creating event handlers in a dedicated directory under src/programs.

hasanlatch007 commented:
Aha then i would need to understand the new event system will ask Adrian for some help :)

hasanlatch007 commented:
Blocked until #403 is merged, DB is bumped with support for user data table, and Winston loggers are added

hasanlatch007 commented:
I moved the `lostRole` and `gainedRole` from `moderation.ts` to `tools.ts` since I feel like it's more suitable there?

geisterfurz007 commented on src/common/moderator.ts:
```diff
@@ -41,3 +42,13 @@ export const getMember = (userId: string): GuildMember => {
   return bot.guilds.resolve(process.env.GUILD_ID).members.resolve(userId);
 };
+
+export const isUserTimedOut = async (
+  member: GuildMember | PartialGuildMember
+): Promise<boolean> => {
+  return !!(await prisma.timedOutUsers.findFirst({
+    where: {
+      userId: member.id,
+    },
+  }));
+};
```
This is specific to the timeout evasion functionality so it should be moved to that file.

geisterfurz007 commented on src/programs/timeout-evasion.ts:
```diff
@@ -0,0 +1,74 @@
+import Tools from "../common/tools";
+import prisma from "../prisma";
+
+@Command({
+  event: DiscordEvent.GUILD_MEMBER_UPDATE,
+  description:
+    "This handle checks if the user was given or removed the Time Out role",
+})
+class TimeoutRoleStatus
+  implements CommandHandler<DiscordEvent.GUILD_MEMBER_UPDATE>
+{
+  async handle(
+    oldMember: GuildMember | PartialGuildMember,
+    newMember: GuildMember | PartialGuildMember
+  ): Promise<void> {
+    if (await Tools.lostRole(oldMember, newMember, "Time Out")) {
```
You can split this a little bit further by creating two classes; one handling the role being added and one handling the role being removed. The options for GUILD_MEMBER_UPDATE take `roleNamesAdded` and `roleNamesRemoved`, so one handler would be

```ts
@Command({
  event: DiscordEvent.GUILD_MEMBER_UPDATE,
  roleNamesAdded: ["Time Out"],
})
```
and the other one the same with `roleNamesRemoved` instead of `roleNamesAdded` (feel free to add descriptions as desired).

That also removes the need for the checks whether the member has gained or lost a role since that is guaranteed to have happened based on the options given.

geisterfurz007 commented on src/programs/timeout-evasion.ts:
```diff
@@ -0,0 +1,74 @@
+            userId: newMember.id,
+          },
+        });
+      } catch (e) {
+        await textLog(
+          `Failed to remove <@${newMember.id}> from timed out DB, please contact a @Developer`
+        );
+      }
```
In such a case, also log the error. Otherwise while a developer may have been contacted, they have absolutely no idea what happened since there is no trace of this event anywhere in the logs.

hasanlatch007 replied:
I did say, ill be adding the logger at the end, just right now it's the functionality I'm aiming for.

geisterfurz007 replied:
Ooooooooooh derp; I misread that comment, sorry!

hasanlatch007 replied:
It's okay we all misread comments sometimes :p

geisterfurz007 commented on src/programs/timeout-evasion.ts:
```diff
@@ -0,0 +1,74 @@
+            userId: newMember.id,
+          },
+        });
+      } catch (e) {
+        await textLog(
+          `Failed to register <@${newMember.id}> to the DB, please contact a @Developer`
+        );
```
See comment above.

geisterfurz007 commented on src/programs/timeout-evasion.ts:
```diff
@@ -0,0 +1,74 @@
+class ReportUserOnJoin implements CommandHandler<DiscordEvent.MEMBER_JOIN> {
+  async handle(member: GuildMember): Promise<void> {
+    if (!(await isUserTimedOut(member))) return;
+    const guild = bot.guilds.resolve(process.env.GUILD_ID);
+    const timeoutRole = Tools.getRoleByName("Time Out", guild);
```
```suggestion
    const timeoutRole = Tools.getRoleByName("Time Out", member.guild);
```

hasanlatch007 commented:
Did some final cleanup, moved all functions necessary, and added the logger, all that is left is the DB, Adrian said he's working on that but since he is busy maybe I can do it? Or at least the part we need.

geisterfurz007 commented:
Database has been bumped to 0.13.0 on master; if you rebase and run yarn, the table should exist. I removed the blocked label.

geisterfurz007 commented:
To have working code on this branch, please rebase this branch:

```
git fetch
git rebase origin/master
# Fix potential merge conflicts and git rebase --continue after fixing them, keep going until the rebase was successful
git push --force-with-lease
```

hasanlatch007 commented:
Should be good to go!

geisterfurz007 commented on src/programs/timeout-evasion.ts:
```diff
@@ -0,0 +1,102 @@
+  async handle(member: GuildMember): Promise<void> {
+    if (await isUserTimedOut(member)) return;
+
+    const engineerRole = Tools.getRoleByName(
+      process.env.ENGINEER_ROLE_NAME,
+      member.guild
+    );
```
You can move this to the catch clause to avoid iterating through the roles in case there was no error.

geisterfurz007 commented on src/programs/timeout-evasion.ts:
```diff
@@ -0,0 +1,102 @@
+        },
+      });
+    } catch (e) {
+      logger.error("Failed to add user to the DB " + e);
```
If `e` is an object, this will end up as `"Failed to add user to the DB [object object]"`. You can pass the error as second argument to have it properly show in the log output:

```suggestion
      logger.error("Failed to add user to the DB ", e);
```

geisterfurz007 commented on src/programs/timeout-evasion.ts:
```diff
@@ -0,0 +1,102 @@
+  implements CommandHandler<DiscordEvent.GUILD_MEMBER_UPDATE>
+{
+  async handle(member: GuildMember): Promise<void> {
+    const engineerRole = Tools.getRoleByName(
+      process.env.ENGINEER_ROLE_NAME,
+      member.guild
+    );
```
See comment above.

geisterfurz007 commented on src/programs/timeout-evasion.ts:
```diff
@@ -0,0 +1,102 @@
+        },
+      });
+    } catch (e) {
+      logger.error("Failed to remove user from the DB " + e);
```
See comment above.
